lib/cstruct.ml

```diff
@@ -282,13 +283,923 @@ let get_uint64 swap p t i =
     let r = ba_get_int64 t.buffer (t.off+i) in
     if swap then swap64 r else r [@@inline]
 
+(* o - bitoffset, i - count of bits *)
+(* It assumes we use __attribute__((packed)) with alignment = 1 byte *)
+(* Here we also assume there is no trailing padding at the end *)
+(* The offset and alignment rules depend on the "mother" type *)
+let align = 8 (* bits, not bytes *)
+
+(* Count the granulas required to read/write - little endian *)
+let bitfield_granula_count_le align _ o i =
+    if o mod align = 0 then (
+        if i mod align = 0 then i / align
+        else i / align + 1
+    ) else (
+        if (i + o) mod align = 0 then i / align
+        else i / align + 1
+    )
+
+(* Count the granulas required to read/write - big endian *)
+let bitfield_granula_count_be align t o i =
+    let o' = t.len * 8 - o - i in
+    if o' mod align = 0 then (
+        if i mod align = 0 then i / align
+        else i / align + 1
+    ) else (
+        if (i + o') mod align = 0 then i / align
+        else i / align + 1
+    )
+
+
+(* These functions do not cross the boundary, and just add zeroes *)
+let get_uint16_chop_boundary swap p t i =
+    if i + 2 > t.len then
+        let part = get_uint8 t i in
+        if swap then part lsl 8 else part
+    else
+        get_uint16 swap p t i
+
+let get_uint32_chop_boundary swap p t i =
+    if i + 4 > t.len then
```

Comment:
I think that these branches could be refactored in to a `match i - t.len with ...` (or `match () with _ when cond1 -> ... | _ when cond2 -> ...`  which would make all branches at the same level)